### PR TITLE
(#1251) Add limit parameter to Split

### PR DESCRIPTION
**Pull request by victornoel**
This is for #1251

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1263](https://github.com/yegor256/cactoos/issues/1263) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#1263](https://github.com/yegor256/cactoos/issues/1263) is assigned to @iakunin[/z](https://www.0crat.com/u/iakunin), here is [why](https://www.0crat.com/footprint/C63314D6Z/8db7ba65-3e3b-4e6a-a873-ba0c7276f167); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

**Comment by paulodamaso** [comment]
@victornoel ping

**Comment by victornoel** [comment]
@paulodamaso @iakunin I haven't forgotten, sorry for the delay, I will get back to it before end of week

**Comment by victornoel** [comment]
@iakunin @paulodamaso sorry for the delay, it's done

**Comment by paulodamaso** [comment]
@victornoel Build is failing, please take a look

**Comment by victornoel** [comment]
@paulodamaso yes, sorry about that, it's fixed :)

**Comment by paulodamaso** [comment]
@iakunin please go on with the review

**Comment by victornoel** [comment]
@iakunin ping

**Comment by victornoel** [comment]
@iakunin ping ?

File: src/main/java/org/cactoos/text/Split.java [hunk 1]
```diff
@@ -45,6 +45,11 @@
      */
     private final UncheckedText regex;
 
+    /**
+     * The limit.
```
Comment (paulodamaso) [inline src/main/java/org/cactoos/text/Split.java]: @victornoel We need more documentation on how this `limit` works: how is the expected behavior with different limit values (> 0, == 0 and <0)

File: src/test/java/org/cactoos/text/SplitTest.java [hunk 2]
```diff
@@ -92,4 +97,57 @@ public void splitTextWithTextRegex() throws Exception {
             new TextHasString("Split")
         ).affirm();
     }
+
```
Comment (paulodamaso) [inline src/test/java/org/cactoos/text/SplitTest.java]: @victornoel PLease add test cases for negative and zero limit values

**Comment by victornoel** [comment]
@paulodamaso thx, it is done

**Comment by victornoel** [comment]
@paulodamaso sorry, there was still some errors, now it is fixed

File: src/test/java/org/cactoos/text/SplitTest.java [hunk 3]
```diff
@@ -23,17 +23,28 @@
...
 /**
  * Test case for {@link Split}.
  * @since 0.9
  * @checkstyle JavadocMethodCheck (500 lines)
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
  */
+@SuppressWarnings(
+    {
+        "PMD.TooManyMethods",
+        "PMD.AvoidDuplicateLiterals"
```
Comment (paulodamaso) [inline src/test/java/org/cactoos/text/SplitTest.java]: @victornoel Make these duplicated literals constants and remove this supression

File: src/test/java/org/cactoos/text/SplitTest.java [hunk 4]
```diff
@@ -92,4 +103,99 @@ public void splitTextWithTextRegex() throws Exception {
...
+            new ScalarHasValue<>(new TextIs("and:foo"))
+        ).affirm();
+    }
+
+    @Test
+    public void splitTextNegativeLimitItemValue() throws Exception {
+        new Assertion<>(
+            "Must split a text with correct item",
+            new ItemAt<>(1, new Split("bar:and:foo", ":", -1)),
+            new ScalarHasValue<>(new TextIs("and"))
```
Comment (paulodamaso) [inline src/test/java/org/cactoos/text/SplitTest.java]: @victornoel wouldn't be better if we checked the exact value for this? What would be the expected result for `limit = -1`?

File: src/test/java/org/cactoos/text/SplitTest.java [hunk 5]
```diff
@@ -92,4 +103,99 @@ public void splitTextWithTextRegex() throws Exception {
...
+        ).affirm();
+    }
+
+    @Test
+    public void splitTextZeroLimitItemValue() throws Exception {
+        new Assertion<>(
+            "Must split a text with correct item",
+            new ItemAt<>(1, new Split("bar:and:foo", ":", 0)),
```
Comment (paulodamaso) [inline src/test/java/org/cactoos/text/SplitTest.java]: @victornoel PLease look at the above

File: src/test/java/org/cactoos/text/SplitTest.java [hunk 6]
```diff
@@ -92,4 +103,99 @@ public void splitTextWithTextRegex() throws Exception {
...
+            new IsEqual<>(2)
+        ).affirm();
+    }
+
+    @Test
+    public void splitTextLengthLimitWithTextItemAndRegex() throws Exception {
+        new Assertion<>(
+            "Must split a text with limited length and text regex",
+            new LengthOf(
+                new Split(new TextOf("fo:and:bo"), new TextOf(":"), 2)
```
Comment (paulodamaso) [inline src/test/java/org/cactoos/text/SplitTest.java]: @victornoel It is better to check for the elements, not the size of the array

File: src/test/java/org/cactoos/text/SplitTest.java (hunk 3 above)
Comment (victornoel) [reply src/test/java/org/cactoos/text/SplitTest.java]: @paulodamaso I find that kind of strange, we do want independent tests: using constants is going against that and also the data used for tests is going to be declared far away frmo the tests. If you don't want the suppression, I will simply use different literals to avoid it. WDYT?

File: src/test/java/org/cactoos/text/SplitTest.java (hunk 4 above)
Comment (victornoel) [reply src/test/java/org/cactoos/text/SplitTest.java]: @paulodamaso I agree, I duplicated existing tests and didn't think to much about it. I will rewrite all the tests

File: src/test/java/org/cactoos/text/SplitTest.java (hunk 6 above)
Comment (victornoel) [reply src/test/java/org/cactoos/text/SplitTest.java]: @paulodamaso I agree, I duplicated existing tests and didn't think to much about it. I will rewrite all the tests

**Comment by victornoel** [comment]
@paulodamaso I've taken your feedback into account, I will take care of it soon. As for the documentation, I put it a the head of the class (I delegated to `String.split` documentation to be precise)

**Comment by victornoel** [comment]
@paulodamaso I've fixed the PR:
- rewrote the tests to be more complete in what they tests
- used `IterableEnvelope` in `Split` in order to have delegation of all the methods of the interface (I needed equals in particular)
- added doc on the constructors
- removed duplicate strings

**Comment by paulodamaso** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job [#1263](https://github.com/yegor256/cactoos/issues/1263) is not in the agenda of @iakunin[/z](https://www.0crat.com/u/iakunin), can't retrieve data and time of add

**Comment by 0crat** [comment]
Job [#1263](https://github.com/yegor256/cactoos/issues/1263) is not in the agenda of @iakunin[/z](https://www.0crat.com/u/iakunin), can't inspect

**Comment by 0crat** [comment]
Code review was too long (58 days), architects (@paulodamaso) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @iakunin[/z](https://www.0crat.com/u/iakunin), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by sereshqua** [comment]
@iakunin please make sure you will find at least 3 issues during next CR

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso)

**Comment by iakunin** [comment]
@paulodamaso @victornoel sorry, guys, I’ve completely forgotten about this review, cause it’s been excluded from my Zerocracy agenda.

**Comment by iakunin** [comment]
> @iakunin please make sure you will find at least 3 issues during next CR

@sereshqua got it.

**Comment by sereshqua** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Quality review completed: +4 point(s) just awarded to @sereshqua[/z](https://www.0crat.com/u/sereshqua)

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 point(s) just awarded to @iakunin[/z](https://www.0crat.com/u/iakunin)
